docs/design/interpreter/exception-handling.md

```diff
@@ -0,0 +1,18 @@
+* Stack walking
+* Exception clauses enumeration
+* Exception handlers invocation
+* Resuming execution after a catch handler exits
+
+The stack walking changes are described in the [stack walking](stackwalk.md) document.
+Exception clauses enumeration in CoreCLR is agnostic of the actual low level details of EH clauses information storage. It uses an interface to a JIT manager that does the real work. As described in the stack walking document, a new InterpreterJitManager will be implemented for the interpreted code.
+To enumerate the EH clauses, it will use data provided by the InterpMethod::clauses/num_clauses.
+
+Regarding the exception handlers invocation, CoreCLR uses native helpers CallCatchFunclet, CallFinallyFunclet and CallFilterFunclet. These will need to be modified to recognize interpreted frames and call the related interpreter functions. CallFinallyFunclet would call interp_run_finally and CallFilterFunclet would call the interp_run_filter. As for the CallCatchFunclet, Mono handles calling catch funclets by calling interp_set_resume_state to record the catch handler IR code address and then returning to the mono_interp_exec_method that checks for this stored state and redirects execution to the catch handler if it is set. In CoreCLR, we may want to handle it differently, in a manner close to how interp_run_finally works.
+
+Resuming execution in the parent of a catch handler after the catch handler exits depends on whether the catch handler is in an interpreted code or in compiled managed code. 
+For the compiled code case, the existing CoreCLR EH code will handle it without changes. It would just restore the context at the resume location.
+For the interpreted code though, CoreCLR will resume execution in the mono_interp_exec_method, then pop some InterpFrame instances from the local linked list (depending on how many interpreted managed frames need to be removed) and restore the interpreter SP from the last popped one and the interpreter IP will be set to the resume location IP.
+
+WASM doesn't support stack unwinding and context manipulation, so the resuming mechanism cannot use context restoring. Mono currently returns from the mono_handle_exception after the catch is executed back to the interp_throw. When the resume frame is in the interpreted frames belonging to the current mono_interp_exec_method, it uses the mechanism described in the previous paragraph to "restore" to the resume context. But when the resume frame is above all the frames belonging to the current mono_interp_exec_method, it exits from the mono_interp_exec_method and then throws a C++ exception (of int32 * type set to NULL) that will propagate through native frames until it is caught in a compiled managed code or in another interpreter function up the call chain (see usage of mono_llvm_catch_exception) where the propagation through the interpreted frames continues the same way as in the previous interpreted frames block.
```

Comment:
How is C++ exception handling implemented in Wasm? Does Wasm has any primitives for stack unwinding to support C++ exception handling?

Reply:
WASM has exception handling support, see https://github.com/WebAssembly/exception-handling/tree/main/proposals/exception-handling

Reply:
Yes, it adds new instructions, new tags section in the wasm format and modifies few existing sections.

Note that the current wasm EH, which is implemented in most browser and engines, is deprecated. The new EH with the `exnref` type is being standardized and is becoming available or included as preview feature. Here you can see the feature extensions table https://webassembly.org/features/

Reply:
All of our current Wasm runtimes also support targeting VMs with no exception extensions.  This is one of those cases where it is useful to think of the browser and WASI separately because the all the major browsers currently support the deprecated Wasm exception instructions and none of the WASI runtimes do.

Reply:
To answer the unwinding part of the question, the stack unwinding is done by the VM itself and not by the running code. So after exception is thrown, the stack is unwound by VM and control flow is transferred to the catch instruction.

On wasm, the clang/llvm handles C++ exceptions by inserting a call to a wrapper in the libunwind after the catch instruction. The wrapper then calls the libcxxabi's personality function. More details can be found in https://llvm.org/docs/ExceptionHandling.html#overview and in the comments on top of https://github.com/dotnet/llvm-project/blob/dotnet/main-19.x/llvm/lib/CodeGen/WasmEHPrepare.cpp